mkosi/run.py

```diff
@@ -359,101 +403,91 @@ def bwrap_cmd(
-            # tools might not work anymore when /usr is replaced wholesale. We also make sure that both
-            # /usr/bin and /usr/sbin/ are searched so that e.g. if the host is Arch and the root is Debian we
-            # don't ignore the binaries from /usr/sbin in the Debian root.
-            env = dict(PATH="/usr/bin:/usr/sbin") | env
-
+    with bwrap_cmd(tools=tools, apivfs=apivfs, scripts=scripts) as bwrap:
         try:
-            result = run([*bwrap, *cmd], text=True, env=env, log=False, **kwargs)
+            result = run(
+                [*bwrap, *cmd],
+                text=True,
+                env=env,
+                log=False,
+                stdin=stdin,
+                stdout=stdout,
+                stderr=stderr,
+                input=input,
+                check=check,
+            )
         except subprocess.CalledProcessError as e:
             if log:
                 logging.error(f"\"{' '.join(str(s) for s in cmd)}\" returned non-zero exit code {e.returncode}.")
             if ARG_DEBUG_SHELL.get():
-                run([*bwrap, "sh"], stdin=sys.stdin, check=False, env=env, log=False)
+                run(
+                    [*bwrap, "sh"],
+                    stdin=sys.stdin,
+                    check=False,
+                    env=env,
+                    log=False,
+                )
             raise e
 
         return result
 
 
-def run_workspace_command(
-    root: Path,
-    cmd: Sequence[PathString],
-    bwrap_params: Sequence[PathString] = (),
-    network: bool = False,
-    stdout: _FILE = None,
-    env: Mapping[str, PathString] = {},
-) -> CompletedProcess:
+def chroot_cmd(root: Path, *, options: Sequence[PathString] = (), network: bool = False) -> Sequence[PathString]:
     cmdline: list[PathString] = [
         "bwrap",
         "--unshare-ipc",
         "--unshare-pid",
         "--unshare-cgroup",
-        "--bind", root, "/",
-        "--tmpfs", "/run",
-        "--tmpfs", "/tmp",
-        "--dev", "/dev",
-        "--proc", "/proc",
-        "--ro-bind", "/sys", "/sys",
+        "--dev-bind", root, "/",
         "--die-with-parent",
-        *bwrap_params,
+        "--setenv", "container", "mkosi",
+        "--setenv", "SYSTEMD_OFFLINE", str(int(network)),
+        "--setenv", "HOME", "/",
+        "--setenv", "PATH", "/usr/bin:/usr/sbin",
+        *options,
     ]
 
-    resolve = root.joinpath("etc/resolv.conf")
-
-    tmp = Path(tempfile.NamedTemporaryFile(delete=False).name)
-    tmp.unlink()
-
     if network:
-        # Bubblewrap does not mount over symlinks and /etc/resolv.conf might be a symlink. Deal with this by
-        # temporarily moving the file somewhere else.
-        if resolve.is_symlink():
-            shutil.move(resolve, tmp)
+        resolve = Path("etc/resolv.conf")
+        if (root / resolve).is_symlink():
+            # bubblewrap cannot mount over symlinks so handle that manually here.
+            resolve = (root / resolve).readlink()
```

Comment:
The idea here is that since overmounting is not possible, we just exchange the target? If so, the comment could be a bit clearer.